src/oop/creating_objects_from_classes/hard/q1/MainTest.java

```diff
@@ -1,59 +1,134 @@
+        String missingStatic = "Your main method is missing the static modifier.";
+        String wrongReturnType = "Your main method does not have the correct return type.";
+        Class<?>[] mainArgs = new Class[]{String[].class};
+        /* Check Method Definitions */
+        assertTrue(testDog.hasMethod("main", mainArgs), String.format(missingMain, testDogName, "main"));
+        assertTrue(testDog.hasModifier("main", mainArgs, "public"), incorrectVisibility);
+        assertTrue(testDog.hasModifier("main", mainArgs, "static"), missingStatic);
+        assertTrue(testDog.hasReturnType("main", mainArgs, Void.TYPE), wrongReturnType);
+
+        String incorrectOutput = "Your main method in the %s class does not print the correct output. Make sure that you call the %s method.";
+        Clause[] clauses = new Clause[]{
+                new StringLiteral("Bark\\!"),
+                new Optional(new NewLine())
         };
-        Object makePetOutput = petClass.callMethod(methodName, arguments, petInstance);
-        assertTrue(makePetOutput instanceof Dog, "Your " + methodName + " method does not return the correct type.");
-        _assertEquals(name, animalClass.getFieldValue(makePetOutput, animalStringFieldName),
-                "Your " + methodName + " method does not return the correct output.");
+        Object[][] mainParameters = {{new String[0], String[].class}};
+        /* Call method and check printed output */
+        testDog.callMethod("main", mainParameters, clauses, String.format(incorrectOutput, testDogName, noiseMethodName));
+
+        /* Check that the constructor was called and with the correct values */
+        String noConstructorCall = "Your main method did not initialize a %s object.";
+        String incorrectValue = "Your main method did not initialize a %s with the correct %s value.";
+        assertTrue(Dog.getConstructorCalled(), String.format(noConstructorCall, objectName));
+        assertEquals(expectedDogName, Dog.getConstructorName(), String.format(incorrectValue, objectName, name));
+        assertEquals(expectedDogAge, Dog.getConstructorAge(), String.format(incorrectValue, objectName, age));
+        assertEquals(expectedDogWeight, Dog.getConstructorWeight(), 0.001, String.format(incorrectValue, objectName, weight));
+
+        /* Check that methods were called */
+        String methodNotCalled = "Your main method did not call the %s method.";
+        assertTrue(Dog.getBarkCalled(), String.format(methodNotCalled, noiseMethodName));
+        assertTrue(Dog.getEatCalled(), String.format(methodNotCalled, consumeMethodName));
+    }
+}
+
+class Dog {
+    private String name;
+    private int age;
+    private double weight;
+    private static Dog dog;
+    private static boolean constructorCalled = false;
+    private static String constructorName;
+    private static double constructorWeight;
+    private static int constructorAge;
```

Comment:
Make those final and leave comment about purpose

Reply:
You can't do this because these values get changed in the course of the test, so they can't be final